src/vm_pages.rs

```diff
@@ -1751,6 +1750,93 @@ impl<'a, T: GuestStagePagingMode> FinalizedVmPages<'a, T> {
         Ok(())
     }
 
+    /// Invalidates a page range.
+    pub fn block_pages(&self, page_addr: GuestPageAddr, len: u64) -> Result<()> {
+        let invalidated = self
+            .inner
+            .root
+            .invalidate_range(page_addr, len, |addr| {
+                self.inner
+                    .page_tracker
+                    .is_mapped_page(addr, self.inner.page_owner_id, MemType::Ram)
+                    || (self.inner.page_tracker.is_shared_page(addr, MemType::Ram)
+                        && !self
+                            .inner
+                            .page_tracker
+                            .is_owned(addr, self.inner.page_owner_id))
```

Comment:
kind of pre-existing and not something that needs to be fixed in this PR, but this logic should probably be in a single `PageTracker` call to avoid bouncing of that lock so much.